src/arch/arm/armlib/entry.S

```diff
@@ -8,60 +8,80 @@
 
 #include <asm/modes.h>
 
-.data
-this_spsr_lr:
-	.word 0x0
-	.word 0x0
-
 .text
 .align 4
 .global	swi_handler
 .global	irq_handler
 
 irq_handler:
-	sub	LR, LR, #4
+	sub    LR, LR, #4
 
-	stmfd	sp!, {r0, r1}
-	ldr	r0, =this_spsr_lr
-	mrs	r1, SPSR
-	stmia	r0, {r1, lr}
-	ldmfd	sp!, {r0, r1}
+	/* save temp register*/
+	stmfd  sp!, {r0}
+	/* storing LP */
+	stmfd  sp!, {lr}
+	/*storing SPSR and valuable regs*/
+	mrs    lr, SPSR
+	/* storing SPSR */
+	stmfd  sp!, {lr}
+	/* pointer to saved on special stack registers */
+	mov    r0, sp
+	/* return stack pointer to normal value */
+	add    sp, sp, #(4 * 3)
 
-	msr	CPSR, #ARM_MODE_SYS | I_BIT | F_BIT
-	stmfd	sp!, {r0-r12, lr}
+	/* return to previous CPU mode with disabled interrupts */
+	//orr    lr, lr, #I_BIT | F_BIT
+	//msr    CPSR, lr
+	msr    CPSR, #ARM_MODE_SYS | I_BIT | F_BIT
 
-	ldr	r0, =this_spsr_lr
-	ldmia	r0, {r1, r2}
-	stmfd	sp!,{r0, r1, r2}
+	/* now we have previous mode and sp */
+	stmfd  sp!, {r1-r12, lr}
 
-	ldr	lr, =irq_leave
-	msr	CPSR, #ARM_MODE_IRQ | I_BIT | F_BIT
+	/* restore spsr, lr, tmp reg */
+	ldmfd  r0!, {r1} /* spsr */
```

Comment:
This starts to load in decreasing addressed. After load stack will be 
```
[r0]
[lr]
[SPSR] 
... <- r0 points here
```

I suppose `ldmia r0, {r1, r2, r3}` should be used.